benchmark/framework/containerd_utils.go

```diff
@@ -0,0 +1,312 @@
+	stdoutPipeReader, stdoutPipeWriter := io.Pipe()
+	stderrPipeReader, stderrPipeWriter := io.Pipe()
+	cioCreator := cio.NewCreator(cio.WithStreams(os.Stdin, stdoutPipeWriter, stderrPipeWriter))
+	task, err := container.NewTask(proc.Context, cioCreator)
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanupFunc := func() {
+		stdoutPipeReader.Close()
+		stdoutPipeWriter.Close()
+		stderrPipeReader.Close()
+		stderrPipeWriter.Close()
+		processStatus, _ := task.Status(proc.Context)
+		if processStatus.Status != "stopped" {
+			fmt.Printf("Tried to kill task")
+			err = task.Kill(proc.Context, syscall.SIGKILL)
+			if err != nil {
+				fmt.Printf("Error killing task: %v\n", err)
+			}
+		}
+		// stat, err := task.Status(proc.Context)
+		// fmt.Printf("task status before delete %v\n", stat)
+		_, err = task.Delete(proc.Context)
+		if err != nil {
+			fmt.Printf("Error deleting task: %v\n", err)
+		}
+	}
+
+	return &TaskDetails{
+			task,
+			stdoutPipeReader,
+			stderrPipeReader},
+		cleanupFunc,
+		err
+}
+
+func (proc *ContainerdProcess) RunContainerTaskForReadyLine(
+	taskDetails *TaskDetails,
+	readyLine string) (func(), error) {
+	stdoutScanner := bufio.NewScanner(taskDetails.stdoutReader)
+	stderrScanner := bufio.NewScanner(taskDetails.stderrReader)
+
+	exitStatusC, err := taskDetails.task.Wait(proc.Context)
+	if err != nil {
+		return nil, err
+	}
+	resultChannel := make(chan string, 1)
+	go func() {
+		<-exitStatusC
+		resultChannel <- "PROC_EXIT"
+	}()
+	go func() {
+		continueLoop := true
+		for stderrScanner.Scan() && continueLoop {
+			nextLine := stderrScanner.Text()
+			if strings.Contains(nextLine, readyLine) {
+				resultChannel <- "READYLINE_STDERR"
+			}
+			select {
+			case <-resultChannel:
+				continueLoop = false
+			default:
+				continueLoop = true
+			}
+		}
+	}()
+	go func() {
+		continueLoop := true
+		for stdoutScanner.Scan() && continueLoop {
+			nextLine := stdoutScanner.Text()
+			if strings.Contains(nextLine, readyLine) {
+				resultChannel <- "READYLINE_STDOUT"
+			}
+			select {
+			case <-resultChannel:
+				continueLoop = false
+			default:
+				continueLoop = true
+			}
+		}
+	}()
+	go func() {
+		timeoutSecs := 180
+		timeoutMaxInterval := timeoutSecs * 10
+		timeoutInterval := 0
+		continueLoop := true
+		for timeoutInterval < timeoutMaxInterval && continueLoop {
+			time.Sleep(100 * time.Millisecond)
+			timeoutInterval++
+			select {
+			case <-resultChannel:
+				continueLoop = false
+			default:
+				continueLoop = true
+			}
+		}
+		resultChannel <- "TIMEOUT"
+	}()
+
+	if err := taskDetails.task.Start(proc.Context); err != nil {
+		return nil, err
+	}
+	<-resultChannel
+	cleanupFunc := func() {
+		processStatus, _ := taskDetails.task.Status(proc.Context)
+		if processStatus.Status == "running" {
+			err = taskDetails.task.Kill(proc.Context, syscall.SIGKILL)
+			if err != nil {
+				fmt.Printf("Error killing task: %v\n", err)
+			}
+			exitChannel, _ := taskDetails.task.Wait(proc.Context)
```

Comment:
If I recall I added this because I needed to wait for the process to successfully move out of the running state.  Without the wait we get these errors:
```
Tried to kill taskError deleting task: task must be stopped before deletion: running: failed precondition
Error deleting container: cannot delete running task TEST_RUN_CONTAINER-1668084161373979375: failed precondition
```

The wait ensures that the task is successfully killed before moving forward